### PR TITLE
chore: payment refactor to use global store prices for billing

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

This pr will update the price logic and billing calculation.

* There will be a global storage price (including read and store). It is a median of all sps' price settings. The global storage price will be updated periodically (e.g., every 30 days).
* The fee for bucket will use the global price, including read fee and store fee. 

### Rationale

* make the price more transparent and steady 

### Example

`QueryGlobalSpStorePriceByTime`

### Changes

Notable changes: 
* remove some price queries and add new ones

### CODE REVIEW

**Review by Pythonberg1997 — APPROVED** [review]

File: x/payment/keeper/storage_fee_charge.go [hunk 1]
```diff
@@ -126,15 +126,15 @@ func (k Keeper) applyFrozenUserFlows(ctx sdk.Context, userFlows types.UserFlows,
 	//var frozenRateChanges []types.StreamRecordChange
 	totalActiveRate, totalFrozenRate := sdk.ZeroInt(), sdk.ZeroInt()
 	for _, flowChange := range userFlows.Flows {
-		outFlow := k.GetOutFlow(ctx, sdk.MustAccAddressFromHex(streamRecord.Account), types.OUT_FLOW_STATUS_FROZEN, sdk.MustAccAddressFromHex(flowChange.ToAddress))
+		outFlow := k.GetOutFlow(ctx, sdk.MustAccAddressFromHex(streamRecord.Account), types.OUT_FLOW_STATUS_ACTIVE, sdk.MustAccAddressFromHex(flowChange.ToAddress))
```
Comment (guagualvcha) [inline x/payment/keeper/storage_fee_charge.go:129]: Did not quite get this.

Comment (guagualvcha) [reply x/payment/keeper/storage_fee_charge.go:129]: reject migration when the account is frozen in both start and end tx.

File: x/sp/keeper/msg_server.go [hunk 2]
```diff
@@ -152,10 +152,7 @@ func (k msgServer) CreateStorageProvider(goCtx context.Context, msg *types.MsgCr
 		FreeReadQuota: msg.FreeReadQuota,
 	}
 	k.SetSpStoragePrice(ctx, spStoragePrice)
-	err = k.UpdateSecondarySpStorePrice(ctx)
-	if err != nil {
-		return nil, err
-	}
+	k.SetSpUpdatePriceTimes(ctx, sp.Id, k.GetSpUpdatePriceTimes(ctx, sp.Id)+1)
```
Comment (guagualvcha) [inline x/sp/keeper/msg_server.go:155]: `k.GetSpUpdatePriceTimes(ctx, sp.Id)` seems unnecessary, as it always to be zero here.

**Review by guagualvcha — CHANGES_REQUESTED** [review]
need change the update price time limitation

File: x/payment/keeper/storage_fee_charge.go (hunk 1 above)
Comment (forcodedancing) [reply x/payment/keeper/storage_fee_charge.go:129]: the following if/else is also switched. the new implementation can handle adding flows when the account is frozen.

File: x/sp/keeper/msg_server.go (hunk 2 above)
Comment (forcodedancing) [reply x/sp/keeper/msg_server.go]: remove the `times` record - when we set the `update global price interval`, it means the env is dev/qa, we don't have to limit the times to update sp price.

File: go.mod [hunk 3]
```diff
@@ -177,7 +177,7 @@ replace (
 	github.com/cometbft/cometbft => github.com/bnb-chain/greenfield-cometbft v0.0.2
 	github.com/cometbft/cometbft-db => github.com/bnb-chain/greenfield-cometbft-db v0.8.1-alpha.1
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
-	github.com/cosmos/cosmos-sdk => github.com/bnb-chain/greenfield-cosmos-sdk v0.2.4-alpha.1
+	github.com/cosmos/cosmos-sdk => github.com/bnb-chain/greenfield-cosmos-sdk v0.2.3-alpha.3.0.20230810070304-b0d00febdec4
```
Comment (forcodedancing) [inline go.mod:180]: check this: https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/267

File: x/sp/keeper/msg_server.go [hunk 4]
```diff
@@ -327,13 +333,18 @@ func (k msgServer) UpdateSpStoragePrice(goCtx context.Context, msg *types.MsgUpd
...
-	if err != nil {
-		return nil, errors.Wrapf(err, "update secondary sp store price failed")
-	}
+
 	return &types.MsgUpdateSpStoragePriceResponse{}, nil
 }
 
+func IsLastDaysOfTheMonth(now time.Time, days int) bool {
+	year, month, _ := now.Date()
+	location := now.Location()
```
Comment (guagualvcha) [inline x/sp/keeper/msg_server.go:342]: is it possible that different region get different result for `now.Location()`?

File: x/sp/keeper/msg_server.go [hunk 5]
```diff
@@ -327,13 +333,18 @@ func (k msgServer) UpdateSpStoragePrice(goCtx context.Context, msg *types.MsgUpd
...
+	nextMonth := time.Date(year, month+1, 1, 0, 0, 0, 0, location)
```
Comment (guagualvcha) [inline x/sp/keeper/msg_server.go:343]: day here is `1`, not `0`, can you double confirm this?

File: x/sp/types/params.go [hunk 6]
```diff
@@ -22,6 +22,10 @@ const (
 	DefaultMaintenanceDurationQuota = 21600 // 6 hour
 	// DefaultNumOfLockUpBlocksForMaintenance defines blocks difference which Sp update itself to Maintenance mode is allowed
 	DefaultNumOfLockUpBlocksForMaintenance = 21600
+	// DefaultUpdateGlobalPriceInterval defines the default time duration for updating global storage price
+	DefaultUpdateGlobalPriceInterval uint64 = 0 // 0 means the global price will be updated at the first day of each month
+	// UpdatePriceDisallowedDays defines the days, counting backward from the end of a month, in which sp is not allowed to update its price
+	DefaultUpdatePriceDisallowedDays uint32 = 2
```
Comment (guagualvcha) [inline x/sp/types/params.go:28]: Please notify Dylan that QA env gets a different setting, like `UpdateGlobalPriceInterval` can be 1k.

Comment (forcodedancing) [reply x/sp/types/params.go:28]: sure

File: x/sp/keeper/msg_server.go (hunk 5 above)
Comment (forcodedancing) [reply x/sp/keeper/msg_server.go]: `now` is passed with UTC timezone. I refactor the code to make it clear.

File: x/sp/keeper/msg_server.go (hunk 4 above)
Comment (forcodedancing) [reply x/sp/keeper/msg_server.go]: `now` is passed with UTC timezone. I refactor the code to make it clear.

**Review by guagualvcha — APPROVED** [review]

**Comment by guagualvcha** [comment]
Please resolve the conflicts

**Comment by forcodedancing** [comment]
> Please resolve the conflicts

resolved

**Review by Fynnss — APPROVED** [review]

**Pull request merged**